File: plugin/src/knowledge/providers/reddit.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const REDDIT_API = "https://www.reddit.com";

interface SearchResponse {
  data: {
    children: Array<{
      data: Post;
    }>;
  };
}

interface Post {
  id: string;
  title: string;
  subreddit: string;
  author: string;
  score: number;
  num_comments: number;
  permalink: string;
  selftext?: string;
  url?: string;
  is_self: boolean;
}

export class RedditProvider implements KnowledgeProvider {
  readonly name = "reddit";

  async isAvailable(): Promise<boolean> {
    return true;
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    try {
      // Check if query specifies a subreddit (r/subreddit query)
      if (query.startsWith("r/")) {
        const rest = query.slice(2);
        const spaceIndex = rest.indexOf(" ");
        if (spaceIndex !== -1) {
          const sub = rest.slice(0, spaceIndex);
          const search = rest.slice(spaceIndex + 1);
          const entries = await this.searchSubreddit(sub, search, maxResults);
          return {
            success: true,
            provider: this.name,
            entries,
          };
        }
      }

      const entries = await this.search(query, maxResults);
      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `Reddit lookup failed: ${message}`,
      };
    }
  }

  private formatCount(n: number): string {
    const abs = Math.abs(n);
    if (abs >= 1_000_000) {
      return `${(n / 1_000_000).toFixed(1)}M`;
    } else if (abs >= 1_000) {
      return `${(n / 1_000).toFixed(1)}k`;
    }
    return String(n);
  }

  private postToEntry(post: Post): KnowledgeEntry {
    const score = this.formatCount(post.score);
    const comments = this.formatCount(post.num_comments);

    const lines: string[] = [];
    lines.push(`r/${post.subreddit} | Score: ${score} | Comments: ${comments}`);
    lines.push(`Posted by u/${post.author}`);

    if (post.selftext && post.is_self) {
      const preview = post.selftext.slice(0, 200);
      const truncated = post.selftext.length > 200 ? `${preview}...` : preview;
      if (truncated) lines.push(truncated);
    }

    const permalink = `https://www.reddit.com${post.permalink}`;

    return {
      title: post.title,
      summary: lines.join("\n"),
      url: permalink,
      source: "reddit",
      metadata: {
        id: post.id,
        subreddit: post.subreddit,
        author: post.author,
        score: post.score,
        numComments: post.num_comments,
        isSelf: post.is_self,
        externalUrl: !post.is_self ? post.url : undefined,
      },
    };
  }

  private async search(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const response = await requestUrl({
      url: `${REDDIT_API}/search.json?q=${encodeURIComponent(query)}&sort=relevance&limit=${limit}&type=link`,
      method: "GET",
      headers: {
        "User-Agent": "VaultTree/0.1.0",
      },
      throw: false,
    });

    if (response.status !== 200) {
      throw new Error(`search failed: ${response.status}`);
    }

    const data = response.json as SearchResponse;
    return data.data.children.map((wrapper) => this.postToEntry(wrapper.data));
  }

  private async searchSubreddit(subreddit: string, query: string, limit: number): Promise<KnowledgeEntry[]> {
    const response = await requestUrl({
      url: `${REDDIT_API}/r/${subreddit}/search.json?q=${encodeURIComponent(query)}&restrict_sr=on&sort=relevance&limit=${limit}`,
      method: "GET",
      headers: {
        "User-Agent": "VaultTree/0.1.0",
      },
      throw: false,
    });

    if (response.status === 404) {
      return [];
    }

    if (response.status !== 200) {
      throw new Error(`search failed: ${response.status}`);
    }

    const data = response.json as SearchResponse;
    return data.data.children.map((wrapper) => this.postToEntry(wrapper.data));
  }
}

File: crates/mcp/src/knowledge/reddit.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::collections::HashMap;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const REDDIT_API: &str = "https://www.reddit.com";

pub struct RedditProvider {
    client: Client,
}

#[derive(Debug, Deserialize)]
struct SearchResponse {
    data: SearchData,
}

#[derive(Debug, Deserialize)]
struct SearchData {
    children: Vec<PostWrapper>,
}

#[derive(Debug, Deserialize)]
struct PostWrapper {
    data: Post,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct Post {
    id: String,
    title: String,
    subreddit: String,
    author: String,
    score: i32,
    num_comments: u32,
    permalink: String,
    selftext: Option<String>,
    url: Option<String>,
    created_utc: f64,
    is_self: bool,
}

impl RedditProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
        }
    }

    fn format_count(n: i32) -> String {
        let abs = n.abs();
        if abs >= 1_000_000 {
            format!("{:.1}M", n as f64 / 1_000_000.0)
        } else if abs >= 1_000 {
            format!("{:.1}k", n as f64 / 1_000.0)
        } else {
            n.to_string()
        }
    }

    fn search(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/search.json?q={}&sort=relevance&limit={}&type=link",
            REDDIT_API,
            urlencoding::encode(query),
            limit
        );

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Err(format!("search failed: {}", response.status()));
        }

        let data: SearchResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .data
            .children
            .into_iter()
            .map(|wrapper| {
                let post = wrapper.data;
                let score = Self::format_count(post.score);
                let comments = Self::format_count(post.num_comments as i32);

                let mut lines = Vec::new();
                lines.push(format!("r/{} | Score: {} | Comments: {}", post.subreddit, score, comments));
                lines.push(format!("Posted by u/{}", post.author));

                // Add preview of selftext if available
                if let Some(text) = &post.selftext {
                    if !text.is_empty() && post.is_self {
                        let preview: String = text.chars().take(200).collect();
                        let truncated = if text.len() > 200 {
                            format!("{}...", preview)
                        } else {
                            preview
                        };
                        lines.push(truncated);
                    }
                }

                let permalink = format!("https://www.reddit.com{}", post.permalink);

                let mut metadata = HashMap::new();
                metadata.insert("id".to_string(), serde_json::json!(post.id));
                metadata.insert("subreddit".to_string(), serde_json::json!(post.subreddit));
                metadata.insert("author".to_string(), serde_json::json!(post.author));
                metadata.insert("score".to_string(), serde_json::json!(post.score));
                metadata.insert("numComments".to_string(), serde_json::json!(post.num_comments));
                metadata.insert("isSelf".to_string(), serde_json::json!(post.is_self));
                if let Some(url) = &post.url {
                    if !post.is_self {
                        metadata.insert("externalUrl".to_string(), serde_json::json!(url));
                    }
                }

                KnowledgeEntry {
                    title: post.title,
                    summary: lines.join("\n"),
                    url: Some(permalink),
                    source: "reddit".to_string(),
                    metadata: Some(metadata),
                }
            })
            .collect())
    }

    fn search_subreddit(&self, subreddit: &str, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/r/{}/search.json?q={}&restrict_sr=on&sort=relevance&limit={}",
            REDDIT_API,
            subreddit,
            urlencoding::encode(query),
            limit
        );

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if response.status().as_u16() == 404 {
            return Ok(Vec::new());
        }

        if !response.status().is_success() {
            return Err(format!("search failed: {}", response.status()));
        }

        let data: SearchResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .data
            .children
            .into_iter()
            .map(|wrapper| {
                let post = wrapper.data;
                let score = Self::format_count(post.score);
                let comments = Self::format_count(post.num_comments as i32);

                let mut lines = Vec::new();
                lines.push(format!("Score: {} | Comments: {}", score, comments));
                lines.push(format!("Posted by u/{}", post.author));

                let permalink = format!("https://www.reddit.com{}", post.permalink);

                let mut metadata = HashMap::new();
                metadata.insert("id".to_string(), serde_json::json!(post.id));
                metadata.insert("subreddit".to_string(), serde_json::json!(post.subreddit));
                metadata.insert("author".to_string(), serde_json::json!(post.author));
                metadata.insert("score".to_string(), serde_json::json!(post.score));
                metadata.insert("numComments".to_string(), serde_json::json!(post.num_comments));

                KnowledgeEntry {
                    title: post.title,
                    summary: lines.join("\n"),
                    url: Some(permalink),
                    source: "reddit".to_string(),
                    metadata: Some(metadata),
                }
            })
            .collect())
    }
}

impl Default for RedditProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for RedditProvider {
    fn name(&self) -> &'static str {
        "reddit"
    }

    fn is_available(&self) -> bool {
        true
    }

    fn lookup(&self, query: &str, options: &LookupOptions) -> LookupResult {
        let limit = options.max_results.unwrap_or(5);

        // Check if query specifies a subreddit (r/subreddit query)
        if let Some(rest) = query.strip_prefix("r/") {
            if let Some((sub, search)) = rest.split_once(' ') {
                return match self.search_subreddit(sub, search, limit) {
                    Ok(entries) => LookupResult::success(self.name(), entries),
                    Err(e) => LookupResult::error(self.name(), e),
                };
            }
        }

        match self.search(query, limit) {
            Ok(entries) => LookupResult::success(self.name(), entries),
            Err(e) => LookupResult::error(self.name(), e),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore] // Requires network
    fn reddit_search() {
        let provider = RedditProvider::new();
        let result = provider.lookup("rust programming", &LookupOptions::default());
        assert!(result.success);
        assert!(!result.entries.is_empty());
    }
}
